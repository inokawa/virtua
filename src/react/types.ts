import { CSSProperties, ReactNode } from "react";

export type ViewportComponentAttributes = Pick<
  React.HTMLAttributes<HTMLElement>,
  "className" | "style" | "id" | "role" | "tabIndex" | "onKeyDown" | "onWheel"
> &
  React.AriaAttributes;

export interface CustomContainerComponentProps {
  style: CSSProperties;
  children: ReactNode;
}

export type CustomContainerComponent = React.ForwardRefExoticComponent<
  React.PropsWithoutRef<CustomContainerComponentProps> &
    React.RefAttributes<any>
>;

/**
 * Props of customized item component for {@link Virtualizer} or {@link WindowVirtualizer}.
 */
export interface CustomItemComponentProps {
  style: CSSProperties;
  index: number;
  children: ReactNode;
}

export type CustomItemComponent = React.ForwardRefExoticComponent<
  React.PropsWithoutRef<CustomItemComponentProps> & React.RefAttributes<any>
>;
