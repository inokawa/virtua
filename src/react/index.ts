export { VList } from "./VList.js";
export type { VListProps, VListHandle } from "./VList.js";
export { Virtualizer } from "./Virtualizer.js";
export type { VirtualizerProps, VirtualizerHandle } from "./Virtualizer.js";
export { WindowVirtualizer } from "./WindowVirtualizer.js";
export type {
  WindowVirtualizerProps,
  WindowVirtualizerHandle,
} from "./WindowVirtualizer.js";
export { VGrid as experimental_VGrid } from "./VGrid.js";
export type {
  VGridItemResize,
  VGridProps,
  VGridHandle,
  CustomCellComponent,
  CustomCellComponentProps,
} from "./VGrid.js";
export type * from "./types.js";
