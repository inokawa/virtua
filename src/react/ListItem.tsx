import {
  memo,
  useRef,
  useMemo,
  CSSProperties,
  ReactElement,
  ReactNode,
} from "react";
import { useIsomorphicLayoutEffect } from "./useIsomorphicLayoutEffect";
import { ItemResizeObserver } from "../core/resizer";
import { refKey } from "./utils";
import { isRTLDocument } from "../core/environment";
import { CustomItemComponent } from "./types";
import { NULL } from "../core/utils";

interface ListItemProps {
  _children: ReactNode;
  _resizer: ItemResizeObserver;
  _index: number;
  _offset: number;
  _hide: boolean;
  _as: "div" | CustomItemComponent;
  _isHorizontal: boolean;
  _isSSR: boolean | undefined;
}

/**
 * @internal
 */
export const ListItem = memo(
  ({
    _children: children,
    _resizer: resizer,
    _index: index,
    _offset: offset,
    _hide: hide,
    _as: Element,
    _isHorizontal: isHorizontal,
    _isSSR: isSSR,
  }: ListItemProps): ReactElement => {
    const ref = useRef<HTMLDivElement>(NULL);

    // The index may be changed if elements are inserted to or removed from the start of props.children
    useIsomorphicLayoutEffect(() => resizer(ref[refKey]!, index), [index]);

    const style = useMemo((): CSSProperties => {
      const style: CSSProperties = {
        position: hide && isSSR ? undefined : "absolute",
        [isHorizontal ? "height" : "width"]: "100%",
        [isHorizontal ? "top" : "left"]: 0,
        [isHorizontal ? (isRTLDocument() ? "right" : "left") : "top"]: offset,
        visibility: !hide || isSSR ? "visible" : "hidden",
      };
      if (isHorizontal) {
        style.display = "flex";
      }
      return style;
    }, [offset, hide, isSSR, isHorizontal]);

    if (typeof Element === "string") {
      return (
        <Element ref={ref} style={style}>
          {children}
        </Element>
      );
    } else {
      return (
        <Element ref={ref} style={style} index={index}>
          {children}
        </Element>
      );
    }
  }
);
