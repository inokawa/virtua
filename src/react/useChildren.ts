import { ReactElement, ReactNode, useMemo } from "react";
import { ItemElement, flattenChildren } from "./utils.js";

/**
 * @internal
 */
export const useChildren = <T>(
  children: ReactNode | ((data: T, i: number) => ReactElement),
  data: ArrayLike<T> | undefined
) => {
  return useMemo((): [(i: number) => ItemElement, number] => {
    if (typeof children === "function") {
      return [(i) => children(data![i]!, i), data!.length];
    }
    // Memoize element array
    const _elements = flattenChildren(children);
    return [(i) => _elements[i]!, _elements.length];
  }, [children, data]);
};
