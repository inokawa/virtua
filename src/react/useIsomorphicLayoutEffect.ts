import { useEffect, useLayoutEffect } from "react";
import { isBrowser } from "../core";

/**
 * https://gist.github.com/gaearon/e7d97cdf38a2907924ea12e4ebdf3c85
 * @internal
 */
export const useIsomorphicLayoutEffect = isBrowser
  ? useLayoutEffect
  : useEffect;
