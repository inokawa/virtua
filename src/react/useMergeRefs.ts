import { Ref, MutableRefObject, useCallback } from "react";

/**
 * @internal
 * Merge multiple refs into a single ref callback.
 */
export function useMergeRefs<T>(...refs: Array<Ref<T> | undefined>): Ref<T> {
  return useCallback((instance: T | null) => {
    for (const ref of refs) {
      ref && updateRef(ref, instance);
    }
  }, refs);
}

function updateRef<T>(ref: Ref<T>, instance: T | null): void {
  if (typeof ref === "function") {
    ref(instance);
  } else if (ref) {
    (ref as MutableRefObject<T | null>).current = instance;
  }
}
