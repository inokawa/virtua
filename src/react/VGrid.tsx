import {
  memo,
  useRef,
  useMemo,
  CSSProperties,
  ReactElement,
  forwardRef,
  ReactNode,
  useImperativeHandle,
} from "react";
import {
  ACTION_ITEMS_LENGTH_CHANGE,
  getOverscanedRange,
  createVirtualStore,
  SCROLL_IDLE,
  getScrollSize,
  UPDATE_VIRTUAL_STATE,
} from "../core/store";
import { useIsomorphicLayoutEffect } from "./useIsomorphicLayoutEffect";
import { createGridScroller } from "../core/scroller";
import { refKey } from "./utils";
import { useStatic } from "./useStatic";
import { createGridResizer, GridResizer } from "../core/resizer";
import { ViewportComponentAttributes } from "./types";
import { flushSync } from "react-dom";
import { isRTLDocument } from "../core/environment";
import { useRerender } from "./useRerender";
import { NULL } from "../core/utils";
const genKey = (i: number, j: number) => `${i}-${j}`;

/**
 * Props of customized cell component for {@link VGrid}.
 */
export interface CustomCellComponentProps {
  style: CSSProperties;
  children: ReactNode;
}

export type CustomCellComponent = React.ForwardRefExoticComponent<
  React.PropsWithoutRef<CustomCellComponentProps> & React.RefAttributes<any>
>;

interface CellProps {
  _children: ReactNode;
  _resizer: GridResizer;
  _rowIndex: number;
  _colIndex: number;
  _top: number;
  _left: number;
  _height: number;
  _width: number;
  _hide: boolean;
  _element: "div";
}

const Cell = memo(
  ({
    _children: children,
    _resizer: resizer,
    _rowIndex: rowIndex,
    _colIndex: colIndex,
    _top: top,
    _left: left,
    _height: height,
    _width: width,
    _hide: hide,
    _element: Element,
  }: CellProps): ReactElement => {
    const ref = useRef<HTMLDivElement>(NULL);

    // The index may be changed if elements are inserted to or removed from the start of props.children
    useIsomorphicLayoutEffect(
      () => resizer._observeItem(ref[refKey]!, rowIndex, colIndex),
      [colIndex, rowIndex]
    );

    return (
      <Element
        ref={ref}
        style={useMemo((): CSSProperties => {
          const style: CSSProperties = {
            display: "grid",
            position: "absolute",
            top: top,
            [isRTLDocument() ? "right" : "left"]: left,
            visibility: hide ? "hidden" : "visible",
            minHeight: height,
            minWidth: width,
          };
          return style;
        }, [top, left, width, height, hide])}
      >
        {children}
      </Element>
    );
  }
);

/**
 * Methods of {@link VGrid}.
 */
export interface VGridHandle {
  /**
   * Get current scrollTop.
   */
  readonly scrollTop: number;
  /**
   * Get current scrollLeft.
   */
  readonly scrollLeft: number;
  /**
   * Get current scrollHeight.
   */
  readonly scrollHeight: number;
  /**
   * Get current scrollWidth.
   */
  readonly scrollWidth: number;
  /**
   * Get current offsetHeight.
   */
  readonly viewportHeight: number;
  /**
   * Get current offsetWidth.
   */
  readonly viewportWidth: number;
  /**
   * Scroll to the item specified by index.
   * @param indexX horizontal index of item
   * @param indexY vertical index of item
   */
  scrollToIndex(indexX: number, indexY: number): void;
  /**
   * Scroll to the given offset.
   * @param offsetX offset from left
   * @param offsetY offset from top
   */
  scrollTo(offsetX: number, offsetY: number): void;
  /**
   * Scroll by the given offset.
   * @param offsetX horizontal offset from current position
   * @param offsetY vertical offset from current position
   */
  scrollBy(offsetX: number, offsetY: number): void;
}

/**
 * Props of {@link VGrid}.
 */
export interface VGridProps extends ViewportComponentAttributes {
  /**
   * A function to create elements rendered by this component.
   */
  children: (arg: {
    /**
     * row index of cell
     */
    rowIndex: number;
    /**
     * column index of cell
     */
    colIndex: number;
  }) => ReactNode;
  /**
   * Total row length of grid.
   */
  row: number;
  /**
   * Total column length of grid.
   */
  col: number;
  /**
   * Cell height hint for unmeasured items. It's recommended to specify this prop if item sizes are fixed and known, or much larger than the defaultValue. It will help to reduce scroll jump when items are measured.
   * @defaultValue 40
   */
  cellHeight?: number;
  /**
   * Cell width hint for unmeasured items. It's recommended to specify this prop if item sizes are fixed and known, or much larger than the defaultValue. It will help to reduce scroll jump when items are measured.
   * @defaultValue 100
   */
  cellWidth?: number;
  /**
   * Number of items to render above/below the visible bounds of the grid. Lower value will give better performance but you can increase to avoid showing blank items in fast scrolling.
   * @defaultValue 2
   */
  overscan?: number;
  /**
   * If set, the specified amount of rows will be mounted in the initial rendering regardless of the container size. This prop is mostly for SSR.
   */
  initialRowCount?: number;
  /**
   * If set, the specified amount of cols will be mounted in the initial rendering regardless of the container size. This prop is mostly for SSR.
   */
  initialColCount?: number;
  /**
   * Component or element type for cell element. This component will get {@link CustomCellComponentProps} as props.
   * @defaultValue "div"
   */
  item?: keyof JSX.IntrinsicElements | CustomCellComponent;
}

/**
 * Virtualized grid component. See {@link VGridProps} and {@link VGridHandle}.
 */
export const VGrid = forwardRef<VGridHandle, VGridProps>(
  (
    {
      children,
      row: rowCount,
      col: colCount,
      cellHeight = 40,
      cellWidth = 100,
      overscan = 2,
      initialRowCount,
      initialColCount,
      item: ItemElement = "div",
      style,
      ...attrs
    },
    ref
  ): ReactElement => {
    const [vStore, hStore, resizer, scroller] = useStatic(() => {
      const _vs = createVirtualStore(rowCount, cellHeight, initialRowCount);
      const _hs = createVirtualStore(colCount, cellWidth, initialColCount);
      return [
        _vs,
        _hs,
        createGridResizer(_vs, _hs),
        createGridScroller(_vs, _hs),
      ];
    });
    // The elements length and cached items length are different just after element is added/removed.
    if (rowCount !== vStore._getItemsLength()) {
      vStore._update(ACTION_ITEMS_LENGTH_CHANGE, [rowCount]);
    }
    if (colCount !== hStore._getItemsLength()) {
      hStore._update(ACTION_ITEMS_LENGTH_CHANGE, [colCount]);
    }

    const vRerender = useRerender(vStore);
    const hRerender = useRerender(hStore);

    const [startRowIndex, endRowIndex] = vStore._getRange();
    const [startColIndex, endColIndex] = hStore._getRange();
    const vScrollDirection = vStore._getScrollDirection();
    const hScrollDirection = hStore._getScrollDirection();
    const vJumpCount = vStore._getJumpCount();
    const hJumpCount = hStore._getJumpCount();
    const height = getScrollSize(vStore);
    const width = getScrollSize(hStore);
    const rootRef = useRef<HTMLDivElement>(NULL);

    useIsomorphicLayoutEffect(() => {
      const root = rootRef[refKey]!;
      // store must be subscribed first because others may dispatch update on init depending on implementation
      const unsubscribeVStore = vStore._subscribe(
        UPDATE_VIRTUAL_STATE,
        (sync) => {
          if (sync) {
            flushSync(vRerender);
          } else {
            vRerender();
          }
        }
      );
      const unsubscribeHStore = hStore._subscribe(
        UPDATE_VIRTUAL_STATE,
        (sync) => {
          if (sync) {
            flushSync(hRerender);
          } else {
            hRerender();
          }
        }
      );
      resizer._observeRoot(root);
      scroller._observe(root);
      return () => {
        unsubscribeVStore();
        unsubscribeHStore();
        resizer._dispose();
        scroller._dispose();
      };
    }, []);

    useIsomorphicLayoutEffect(() => {
      scroller._fixScrollJump();
    }, [vJumpCount, hJumpCount]);

    useImperativeHandle(
      ref,
      () => {
        return {
          get scrollTop() {
            return vStore._getScrollOffset();
          },
          get scrollLeft() {
            return hStore._getScrollOffset();
          },
          get scrollHeight() {
            return getScrollSize(vStore);
          },
          get scrollWidth() {
            return getScrollSize(hStore);
          },
          get viewportHeight() {
            return vStore._getViewportSize();
          },
          get viewportWidth() {
            return hStore._getViewportSize();
          },
          scrollToIndex: scroller._scrollToIndex,
          scrollTo: scroller._scrollTo,
          scrollBy: scroller._scrollBy,
        };
      },
      []
    );

    const render = useMemo(() => {
      const cache = new Map<string, ReactNode>();
      return (rowIndex: number, colIndex: number) => {
        let e: ReactNode | undefined = cache.get(genKey(rowIndex, colIndex));
        if (!e) {
          cache.set(
            genKey(rowIndex, colIndex),
            (e = children({ rowIndex, colIndex }))
          );
        }
        return e;
      };
    }, [children]);

    const [overscanedStartRowIndex, overscanedEndRowIndex] = getOverscanedRange(
      startRowIndex,
      endRowIndex,
      overscan,
      vScrollDirection,
      rowCount
    );
    const [overscanedStartColIndex, overscanedEndColIndex] = getOverscanedRange(
      startColIndex,
      endColIndex,
      overscan,
      hScrollDirection,
      colCount
    );

    const items: ReactElement[] = [];
    for (
      let rowIndex = overscanedStartRowIndex;
      rowIndex <= overscanedEndRowIndex;
      rowIndex++
    ) {
      for (
        let colIndex = overscanedStartColIndex;
        colIndex <= overscanedEndColIndex;
        colIndex++
      ) {
        items.push(
          <Cell
            key={genKey(rowIndex, colIndex)}
            _resizer={resizer}
            _rowIndex={rowIndex}
            _colIndex={colIndex}
            _top={vStore._getItemOffset(rowIndex)}
            _left={hStore._getItemOffset(colIndex)}
            _height={vStore._getItemSize(rowIndex)}
            _width={hStore._getItemSize(colIndex)}
            _hide={
              vStore._isUnmeasuredItem(rowIndex) ||
              hStore._isUnmeasuredItem(colIndex)
            }
            _element={ItemElement as "div"}
            _children={render(rowIndex, colIndex)}
          />
        );
      }
    }

    return (
      <div
        ref={rootRef}
        {...attrs}
        style={{
          overflow: "auto",
          contain: "strict",
          width: "100%",
          height: "100%",
          ...style,
        }}
      >
        <div
          style={{
            overflowAnchor: "none", // opt out browser's scroll anchoring because it will conflict to scroll anchoring of virtualizer
            flex: "none", // flex style can break layout
            position: "relative",
            visibility: "hidden", // TODO replace with other optimization methods
            width: width,
            height: height,
            pointerEvents:
              vScrollDirection !== SCROLL_IDLE ||
              hScrollDirection !== SCROLL_IDLE
                ? "none"
                : undefined,
          }}
        >
          {items}
        </div>
      </div>
    );
  }
);
