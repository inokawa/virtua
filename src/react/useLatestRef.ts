import { useRef } from "react";
import { useIsomorphicLayoutEffect } from "./useIsomorphicLayoutEffect.js";
import { refKey } from "./utils.js";

/**
 * @internal
 */
export const useLatestRef = <T>(value: T) => {
  const ref = useRef<T>(value);

  useIsomorphicLayoutEffect(() => {
    ref[refKey] = value;
  }, [value]);

  return ref;
};
