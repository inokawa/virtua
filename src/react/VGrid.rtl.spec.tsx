import { afterEach, it, expect, describe, vi } from "vitest";
import { render as _render, cleanup } from "@testing-library/react";
import { VGrid } from "./VGrid";
import { setupResizeJsDom } from "../../scripts/spec";

Object.defineProperty(CSSStyleDeclaration.prototype, "direction", {
  get() {
    return "rtl";
  },
});

const ITEM_HEIGHT = 50;
const ITEM_WIDTH = 100;
const VIEWPORT_HEIGHT = ITEM_HEIGHT * 10;

setupResizeJsDom({
  itemSize: { width: ITEM_WIDTH, height: ITEM_HEIGHT },
  viewportSize: { width: ITEM_WIDTH, height: VIEWPORT_HEIGHT },
});

const render = (...args: Parameters<typeof _render>) => {
  const res = _render(...args);
  vi.runAllTicks();
  return res;
};

afterEach(cleanup);

describe("rtl", () => {
  it("should work", () => {
    const { asFragment } = render(
      <VGrid row={100} col={100}>
        {({ rowIndex, colIndex }) => (
          <div>
            {rowIndex} / {colIndex}
          </div>
        )}
      </VGrid>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
