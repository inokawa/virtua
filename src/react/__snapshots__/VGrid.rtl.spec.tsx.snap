// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rtl > should work 1`] = `
<DocumentFragment>
  <div
    style="overflow: auto; contain: strict; width: 100%; height: 100%;"
  >
    <div
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 10000px; height: 4040px;"
    >
      <div
        style="display: grid; position: absolute; top: 0px; right: 0px; visibility: visible; min-height: 50px; min-width: 100px;"
      >
        <div>
          0 / 0
        </div>
      </div>
      <div
        style="display: grid; position: absolute; top: 0px; right: 100px; visibility: visible; min-height: 50px; min-width: 100px;"
      >
        <div>
          0 / 1
        </div>
      </div>
      <div
        style="display: grid; position: absolute; top: 0px; right: 200px; visibility: visible; min-height: 50px; min-width: 100px;"
      >
        <div>
          0 / 2
        </div>
      </div>
      <div
        style="display: grid; position: absolute; top: 0px; right: 300px; visibility: visible; min-height: 50px; min-width: 100px;"
      >
        <div>
          0 / 3
        </div>
      </div>
      <div
        style="display: grid; position: absolute; top: 50px; right: 0px; visibility: visible; min-height: 50px; min-width: 100px;"
      >
        <div>
          1 / 0
        </div>
      </div>
      <div
        style="display: grid; position: absolute; top: 50px; right: 100px; visibility: visible; min-height: 50px; min-width: 100px;"
      >
        <div>
          1 / 1
        </div>
      </div>
      <div
        style="display: grid; position: absolute; top: 50px; right: 200px; visibility: visible; min-height: 50px; min-width: 100px;"
      >
        <div>
          1 / 2
        </div>
      </div>
      <div
        style="display: grid; position: absolute; top: 50px; right: 300px; visibility: visible; min-height: 50px; min-width: 100px;"
      >
        <div>
          1 / 3
        </div>
      </div>
      <div
        style="display: grid; position: absolute; top: 100px; right: 0px; visibility: visible; min-height: 50px; min-width: 100px;"
      >
        <div>
          2 / 0
        </div>
      </div>
      <div
        style="display: grid; position: absolute; top: 100px; right: 100px; visibility: visible; min-height: 50px; min-width: 100px;"
      >
        <div>
          2 / 1
        </div>
      </div>
      <div
        style="display: grid; position: absolute; top: 100px; right: 200px; visibility: visible; min-height: 50px; min-width: 100px;"
      >
        <div>
          2 / 2
        </div>
      </div>
      <div
        style="display: grid; position: absolute; top: 100px; right: 300px; visibility: visible; min-height: 50px; min-width: 100px;"
      >
        <div>
          2 / 3
        </div>
      </div>
      <div
        style="display: grid; position: absolute; top: 150px; right: 0px; visibility: visible; min-height: 50px; min-width: 100px;"
      >
        <div>
          3 / 0
        </div>
      </div>
      <div
        style="display: grid; position: absolute; top: 150px; right: 100px; visibility: visible; min-height: 50px; min-width: 100px;"
      >
        <div>
          3 / 1
        </div>
      </div>
      <div
        style="display: grid; position: absolute; top: 150px; right: 200px; visibility: visible; min-height: 50px; min-width: 100px;"
      >
        <div>
          3 / 2
        </div>
      </div>
      <div
        style="display: grid; position: absolute; top: 150px; right: 300px; visibility: visible; min-height: 50px; min-width: 100px;"
      >
        <div>
          3 / 3
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
